From: Michael Schutte <michi@uiae.at>
Subject: [PATCH] fixes/openvt_error_handling

Improved error handling for the openvt program.

This closes the Debian bug #190387.

Signed-off-by: Michael Schutte <michi@uiae.at>

---
 src/openvt.c |   13 +++++++++++--
 1 files changed, 11 insertions(+), 2 deletions(-)

diff --git a/src/openvt.c b/src/openvt.c
index ca8617d..00897c7 100644
--- a/src/openvt.c
+++ b/src/openvt.c
@@ -53,6 +53,7 @@ main(int argc, char *argv[])
    char verbose = FALSE;
    char direct_exec	= FALSE;
    char do_wait	= FALSE;
+   int status;
    char as_user = FALSE;
    char vtname[sizeof VTNAME + 2]; /* allow 999 possible VTs */
    char *cmd = NULL, *def_cmd = NULL, *username = NULL;
@@ -275,7 +276,8 @@ got_vtno:
       close(1);
       close(2);
       close(fd0);
-      close(consfd);
+      if (consfd > 2)
+         close(consfd);
       if ((dup(fd1) == -1) || (dup(fd1) == -1)) {
 	perror("dup");
 	fflush(stderr);
@@ -306,7 +308,9 @@ got_vtno:
    }
 
    if ( do_wait ) {
-      wait(NULL);
+      wait(&status);
+      if (!direct_exec)
+	 close(fd0);
       if (show) { /* Switch back... */
 	 if (ioctl(consfd, VT_ACTIVATE, vtstat.v_active)) {
 	    perror("VT_ACTIVATE");
@@ -323,6 +327,11 @@ got_vtno:
 	   return(8);
 	 }
       }
+
+      if (WIFEXITED(status))
+	 return WEXITSTATUS(status);
+      else
+	 return 127;
    }
 
    return 0;
-- 
tg: (cfffc72..) fixes/openvt_error_handling (depends on: upstream/master)
